pkg/algorithm/nodeselect/v1alpha2/select_node.go

```diff
@@ -0,0 +1,206 @@
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	bd "github.com/openebs/maya/pkg/blockdevice/v1alpha2"
+	bdc "github.com/openebs/maya/pkg/blockdeviceclaim/v1alpha1"
+	csp "github.com/openebs/maya/pkg/cstor/newpool/v1alpha3"
+	"github.com/openebs/maya/pkg/volume"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GetCandidateNodeMap returns a map of all nodes where the pool needs to be created.
+func (ac *Config) GetCandidateNodeMap() (map[string]bool, error) {
+	// TODO : Do not select a node if it is not ready
+	candidateNodesMap := make(map[string]bool)
+	usedNodeMap, err := ac.GetUsedNodeMap()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get candidate nodes for pool creation")
+	}
+	for _, pool := range ac.CSPC.Spec.Pools {
+		nodeName := pool.NodeSelector[HostName]
+		if usedNodeMap[nodeName] == false {
+			candidateNodesMap[nodeName] = true
+		}
+	}
+	return candidateNodesMap, nil
+}
+
+// GetUsedNodeMap returns a map of node for which pool has already been created.
+func (ac *Config) GetUsedNodeMap() (map[string]bool, error) {
+	usedNodeMap := make(map[string]bool)
+	cspList, err := csp.
+		NewKubeClient().
+		WithNamespace(ac.Namespace).
+		List(
+			metav1.
+				ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + ac.CSPC.Name},
+		)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNodeMap[cspObj.Labels[string(apis.HostNameCPK)]] = true
+	}
+	return usedNodeMap, nil
+}
+
+// SelectNode selects a node and returns the pool spec from the cspc for pool provisioning.
+func (ac *Config) SelectNode() (*apis.PoolSpec, error) {
+	candidateNodes, err := ac.GetCandidateNodeMap()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get pool spec for pool creation")
+	}
+	for _, pool := range ac.CSPC.Spec.Pools {
+		pool := pool
+		nodeName := pool.NodeSelector[HostName]
+		if candidateNodes[nodeName] {
+			if ValidatePoolSpec(&pool) {
+				return &pool, nil
+			}
+		}
+	}
+	return nil, errors.New("no node qualified for pool creation")
+}
+
+// GetBDListForNode returns a list of BD from the pool spec.
+// TODO : Move it to CStorPoolCluster packgage
+func (ac *Config) GetBDListForNode(pool *apis.PoolSpec) []string {
+	var BDList []string
+	for _, group := range pool.RaidGroups {
+		for _, bd := range group.BlockDevices {
+			BDList = append(BDList, bd.BlockDeviceName)
+		}
+	}
+	return BDList
+}
+
+// ClaimBDsForNode claims a given BlockDevice for node
+// If the block device(s) is/are already claimed for any other CSPC it returns error.
+// If the block device(s) is/are already calimed for the same CSPC -- it is left as it is and can be used for
+// pool provisioning.
+// If the block device(s) is/are unclaimed, then those are claimed.
+func (ac *Config) ClaimBDsForNode(BD []string) error {
+	for _, bd := range BD {
+		IsBDClaimed, err := ac.IsBDClaimed(bd)
```

Comment:
code seems to be getting BDs multiple times.
It would be good if we can get all BDs at once and use them for all checks